src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,409 @@
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef UTILS_COMMUNICATION_BUFFER_HPP_
+#define UTILS_COMMUNICATION_BUFFER_HPP_
+
+#include <type_traits>
+#include <iostream>
+#include <memory>
+#include <utility>
+#include <unordered_map>
+
+#ifdef MPI_PARALLEL
+#include <mpi.h>
+
+#define request_t MPI_Request
+#define comm_t MPI_Comm
+#else
+#define request_t int
+#define comm_t int
+#endif
+
+namespace parthenon { 
+namespace detail
+{
+
+#ifdef MPI_PARALLEL
+  // MPIType<A> -> MPIType<A, bool_t<true>> and then
+  // specialization checks are performed. This is why the
+  // default parameter is required.
+  template <bool>
+  struct bool_t
+  {
+  };
+
+  template <class T, class = bool_t<true>>
+  struct MPIType;
+
+  template <class U, class T>
+  using MPI_type_check = std::is_same<U, std::remove_pointer_t<T>>;
+
+  template <class T>
+  struct MPIType<T, bool_t<MPI_type_check<double, T>::value>>
+  {
+    MPI_Datatype static value() noexcept { return MPI_DOUBLE; }
+  };
+
+  template <class T>
+  struct MPIType<T, bool_t<MPI_type_check<Real, T>::value>>
+  {
+    MPI_Datatype static value() noexcept { return MPI_PARTHENON_REAL; }
+  };
+
+  template <class T>
+  struct MPIType<T, bool_t<MPI_type_check<int, T>::value>>
+  {
+    MPI_Datatype static value() noexcept { return MPI_INT; }
+  };
+
+  template <class T>
+  struct MPIType<T, bool_t<MPI_type_check<bool, T>::value>>
+  {
+    MPI_Datatype static value() noexcept { return MPI_CXX_BOOL; }
+  };
+#endif
+}
+
+using namespace detail;
+
+class VariableCommMap
+{
+public:
+  // This needs to be called in the same order on every process
+  // TODO: Come up with a more robust scheme for creating variable communicators
+  void AddVariable(const std::string &var_name)
+  {
+    if (map_.count(var_name) > 0)
+      return; // Variable is already in map
+#ifdef MPI_PARALLEL
+    comm_t temp;
+    MPI_Comm_dup(MPI_COMM_WORLD, &temp);
+    map_[var_name] = temp;
+#else
+    map_[var_name] = current_idx_;
+    ++current_idx_;
+#endif
+    return;
+  }
+
+  comm_t GetCommunicator(const std::string &var_name)
+  {
+    return map_[var_name];
+  }
+
+private:
+  comm_t current_idx_; // Only needed when there is no MPI
+  std::unordered_map<std::string, comm_t> map_;
+};
+
+enum class BufferState
+{
+  stale,
+  sending,
+  sending_null,
+  receiving,
+  received,
+  received_null
+};
+
+enum class BuffCommType
+{
+  sender,
+  receiver,
+  both
+};
+
+template <class T>
+class CommBuffer : public T
```

Comment:
I only skimmed this class, but I think I quite like it. It's nice to have this extra machinery.

My one complaint is I dislike using inheritence in this way, as it feels like kind of a bad code smell. I'd rather make `T` a public member.

Reply:
IMHO, the inheritance makes it clearer that a `CommBuffer<T>` *is a* `T` with some extra functionality added and makes user code a little cleaner. But I am not too worried about it.

Reply:
yeah it's not a huge deal either way. I'm just leery of inheritance of things like `stl` containers or `Kokkos::View`s as I think it means you can accidentally carry around functionality you're not aware you're carrying.

Reply:
I think the major issue is that `stl` containers and `Kokkos::View` don't have virtual dtors, so if you try to do something like `Kokkos::View<...>* pview = new CommBuffer<Kokkos::View<...>>(...);` you get undefined behavior when this object is destroyed. I don't think that we are planning on doing things like this, but this is a pretty strong argument for not doing what I am doing here. 

I think this will practically not be an issue for the `Pool` member types, since they contain only to new member fields which are not class types (and therefore don't have dtors) even though technically this is undefined behavior. `CommBuffer<T>` has class fields so I think there is a stronger case to make `T` a member rather than inherited from here.

Reply:
I switched from inheritance to composition for `CommBuffer<T>` but not for `ObjectPool<T>` member types.